File: src/utils/debug.ts
import { DebugOptions } from '../types';

/**
 * Default debug options
 */
const defaultDebugOptions: DebugOptions = {
  enabled: false,
  logRequests: false,
  logResponses: false,
  logger: (message, data) => {
    if (data) {
      console.log(`[@plust/search-sdk] ${message}`, data);
    } else {
      console.log(`[@plust/search-sdk] ${message}`);
    }
  }
};

/**
 * Debug utility for the search SDK
 */
export const debug = {
  /**
   * Log a message if debugging is enabled
   * 
   * @param options Debug options from search request
   * @param message Message to log
   * @param data Optional data to log
   */
  log(options: DebugOptions | undefined, message: string, data?: unknown): void {
    const opts = { ...defaultDebugOptions, ...options };
    if (opts.enabled) {
      const logger = opts.logger || defaultDebugOptions.logger;
      if (logger) {
        logger(message, data);
      }
    }
  },

  /**
   * Log request details if request logging is enabled
   * 
   * @param options Debug options from search request
   * @param message Message to log
   * @param data Request details to log
   */
  logRequest(options: DebugOptions | undefined, message: string, data?: unknown): void {
    const opts = { ...defaultDebugOptions, ...options };
    if (opts.enabled && opts.logRequests) {
      const logger = opts.logger || defaultDebugOptions.logger;
      if (logger) {
        logger(`REQUEST: ${message}`, data);
      }
    }
  },

  /**
   * Log response details if response logging is enabled
   * 
   * @param options Debug options from search request
   * @param message Message to log
   * @param data Response details to log
   */
  logResponse(options: DebugOptions | undefined, message: string, data?: unknown): void {
    const opts = { ...defaultDebugOptions, ...options };
    if (opts.enabled && opts.logResponses) {
      const logger = opts.logger || defaultDebugOptions.logger;
      if (logger) {
        logger(`RESPONSE: ${message}`, data);
      }
    }
  }
};

File: src/utils/debug.rs
//! Debug utilities for the search SDK

use crate::types::DebugOptions;

/// Log a message if debugging is enabled
pub fn log(options: &Option<DebugOptions>, message: &str, data: &str) {
    if let Some(debug_opts) = options {
        if debug_opts.enabled {
            eprintln!("[search-sdk] {message}: {data}");
        }
    }
}

/// Log request details if request logging is enabled
pub fn log_request(options: &Option<DebugOptions>, message: &str, data: &str) {
    if let Some(debug_opts) = options {
        if debug_opts.enabled && debug_opts.log_requests {
            eprintln!("[search-sdk] REQUEST: {message}: {data}");
        }
    }
}

/// Log response details if response logging is enabled
pub fn log_response(options: &Option<DebugOptions>, message: &str) {
    if let Some(debug_opts) = options {
        if debug_opts.enabled && debug_opts.log_responses {
            eprintln!("[search-sdk] RESPONSE: {message}");
        }
    }
}

/// Create default debug options with all logging enabled
pub fn debug_all() -> DebugOptions {
    DebugOptions {
        enabled: true,
        log_requests: true,
        log_responses: true,
    }
}

/// Create debug options with only basic logging enabled
pub fn debug_basic() -> DebugOptions {
    DebugOptions {
        enabled: true,
        log_requests: false,
        log_responses: false,
    }
}
